scikits/odes/sundials/cvode.pyx

```diff
@@ -1180,22 +1205,66 @@ cdef class CVODE:
         np_tspan = np.asarray(tspan, dtype=float)
         np_y0    = np.asarray(y0, dtype=float)
 
-        return self._solve(np_tspan, np_y0)
+        soln = self._solve(np_tspan, np_y0)
+
+        flag_no, t, y, t_err, y_err = soln[:5]
+
+        if soln[5]:
+            t_roots = np.array(soln[5])
+        else:
+            t_roots = None
+
+        if soln[6]:
+            y_roots = np.array(soln[6])
+        else:
+            y_roots = None
+
+        if soln[7]:
+            t_tstop = np.array(soln[7])
+        else:
+            t_tstop = None
+
+        if soln[8]:
+            y_tstop = np.array(soln[8])
+        else:
+            y_tstop = None
+
+        flag = self.RETURN_FLAGS.get(flag_no, "UNKNOWN")
+
+        soln = CVODE_Return(
+            flag=flag, flag_no=flag_no, values=(t,y), err_values=(t_err,y_err),
+            root_values=(t_roots,y_roots), tstop_values=(t_tstop,y_tstop)
+        )
+
```

Comment:
About your new API proposal:
I agree that returning all data in a single object is better than querying the solver object, so yes, we go this way.
1. I'd like to see also the error message (basically your comment in
RETURN_FLAGS) in the tuple. On error I believe that this is the most important information about what happened and one does not need to check the documentation (the error code can be included in the message however).
2. We should have an easy way of checking for the error - in the old API we could use simple ```flag < 0```, in the new one can check for ```soln.error is None```. However the 'flag_name' approach itself does not allow such check so easily (I know that the suggestion is to return both flag_name and flag_number, just pointing this out).
3. When an error occurs, one probably does not need the 'flag_name' names - and if we include the error message, probably also not the 'flag_number'. I may be wrong about this, but I don't assume anyone would continue the computation on (such CVODE) error. We can still pass the flag though.
4. If we exclude the error case, we are left with three (other) possible states - success, tstop and root. And for that we have 'flag_name' and 'flag_number' which provide the same (duplicit) information. We should/could drop one of them completely, probably 'flag_number'.
5. One possible way would be to use IntEnum, which would allow us to use 'flag_name' while not deviating from the CVODE flags: https://docs.python.org/3/library/enum.html#intenum.
6. In case 'interruptfn' is not supplied but e.g. root found (via 'rootfn' supplied), how is the ```(t_root, y_root)``` returned in the new API?